File: content-node/websocket_test.go
package main

// Mirrors packages/protocol/test/websocket.test.ts. That package is the
// normative definition of the wire format; these assert this side agrees.

import (
	"strings"
	"testing"
)

func TestWebSocketDataRoundTrip(t *testing.T) {
	for _, tc := range []struct {
		name string
		kind WebSocketDataKind
		body []byte
	}{
		{"text", WebSocketText, []byte("hello wisp")},
		{"binary", WebSocketBinary, []byte{0, 1, 2, 253, 254, 255}},
		{"empty text", WebSocketText, []byte{}},
		{"empty binary", WebSocketBinary, []byte{}},
	} {
		t.Run(tc.name, func(t *testing.T) {
			payload, err := encodeWebSocketData(tc.kind, tc.body)
			if err != nil {
				t.Fatalf("encode: %v", err)
			}
			kind, data, err := decodeWebSocketData(payload)
			if err != nil {
				t.Fatalf("decode: %v", err)
			}
			if kind != tc.kind {
				t.Fatalf("kind = %d, want %d", kind, tc.kind)
			}
			if string(data) != string(tc.body) {
				t.Fatalf("data = %q, want %q", data, tc.body)
			}
		})
	}
}

func TestWebSocketDataRejectsOversizeMessage(t *testing.T) {
	if _, err := encodeWebSocketData(WebSocketBinary, make([]byte, maxWebSocketMessageBytes)); err != nil {
		t.Fatalf("a message at the ceiling must encode: %v", err)
	}
	if _, err := encodeWebSocketData(WebSocketBinary, make([]byte, maxWebSocketMessageBytes+1)); err == nil {
		t.Fatal("a message past the ceiling must not encode")
	}
}

func TestWebSocketDataRejectsUnknownKind(t *testing.T) {
	if _, _, err := decodeWebSocketData([]byte{7, 1, 2}); err == nil {
		t.Fatal("an unknown kind must be rejected rather than guessed")
	}
	if _, _, err := decodeWebSocketData(nil); err == nil {
		t.Fatal("an empty payload must be rejected")
	}
}

func TestWebSocketCloseRoundTrip(t *testing.T) {
	payload, err := encodeWebSocketClose(1000, "done")
	if err != nil {
		t.Fatalf("encode: %v", err)
	}
	code, reason, err := decodeWebSocketClose(payload)
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	if code != 1000 || reason != "done" {
		t.Fatalf("got %d %q, want 1000 \"done\"", code, reason)
	}
}

func TestWebSocketCloseRejectsLocallySynthesisedCodes(t *testing.T) {
	// 1005 and 1006 mean "no close frame was seen"; sending either would be a
	// lie about what happened.
	for _, code := range []int{1005, 1006, 0, 999, 0x10000} {
		if _, err := encodeWebSocketClose(code, ""); err == nil {
			t.Fatalf("close code %d must be rejected", code)
		}
	}
}

func TestWebSocketCloseRejectsOversizeReason(t *testing.T) {
	if _, err := encodeWebSocketClose(1000, strings.Repeat("x", 124)); err == nil {
		t.Fatal("a reason past 123 bytes must be rejected")
	}
	if _, err := encodeWebSocketClose(1000, strings.Repeat("x", 123)); err != nil {
		t.Fatalf("a reason at 123 bytes must encode: %v", err)
	}
}

func TestWebSocketCloseRejectsTruncatedPayload(t *testing.T) {
	if _, _, err := decodeWebSocketClose([]byte{3}); err == nil {
		t.Fatal("a truncated close payload must be rejected")
	}
}

func TestSanitizeCloseCode(t *testing.T) {
	for _, code := range []int{1005, 1006, 0, 999, 0x10000} {
		if got := sanitizeCloseCode(code); got != wsCloseGoingAway {
			t.Fatalf("sanitizeCloseCode(%d) = %d, want %d", code, got, wsCloseGoingAway)
		}
	}
	if got := sanitizeCloseCode(1000); got != 1000 {
		t.Fatalf("a legal code must survive, got %d", got)
	}
}

// The node must not become a general-purpose dialler for whatever the page asks.
func TestWebSocketTargetRefusesAnythingButTheConfiguredUpstream(t *testing.T) {
	handler := &Handler{WebSocketURL: "ws://127.0.0.1:1802"}

	for _, tc := range []struct {
		name string
		url  string
	}{
		{"absolute url", "https://evil.example/socket"},
		{"scheme-relative url", "//evil.example/socket"},
		{"outside the api namespace", "/socket"},
		{"static path", "/index.html"},
		{"api lookalike prefix", "/apiv2evil/socket"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if _, err := handler.webSocketTarget(tc.url); err == nil {
				t.Fatalf("%s must be refused", tc.url)
			}
		})
	}
}

func TestWebSocketTargetAcceptsTheApiNamespace(t *testing.T) {
	handler := &Handler{WebSocketURL: "ws://127.0.0.1:1802"}

	target, err := handler.webSocketTarget("/apiv2/wonderlands/?t=1")
	if err != nil {
		t.Fatalf("an api path must be accepted: %v", err)
	}
	if target != "ws://127.0.0.1:1802/apiv2/wonderlands/?t=1" {
		t.Fatalf("target = %q", target)
	}
}

func TestWebSocketDisabledByDefault(t *testing.T) {
	handler := &Handler{}
	if _, err := handler.webSocketTarget("/apiv2/wonderlands/"); err == nil {
		t.Fatal("a node with no configured upstream must not open sockets")
	}
}

func TestWebSocketOpenFromFrameBoundsItsInput(t *testing.T) {
	if _, err := webSocketOpenFromFrame(make([]byte, maxWebSocketOpenBytes+1)); err == nil {
		t.Fatal("an oversize open payload must be rejected")
	}
	if _, err := webSocketOpenFromFrame([]byte("not json")); err == nil {
		t.Fatal("a malformed open payload must be rejected")
	}
	if _, err := webSocketOpenFromFrame([]byte(`{"protocols":["a\r\nb"]}`)); err == nil {
		t.Fatal("a subprotocol carrying header separators must be rejected")
	}
	if _, err := webSocketOpenFromFrame([]byte(`{"version":3,"url":"/apiv2/x","protocols":[],"initialCredits":16}`)); err != nil {
		t.Fatalf("a well-formed open must parse: %v", err)
	}
}

// The frame types must decode, not merely have payload codecs. knownType is a
// range, so a new type that is not inside it is dropped before dispatch ever
// sees it -- which looks exactly like the node ignoring the peer.
func TestWebSocketFrameTypesDecode(t *testing.T) {
	for _, frameType := range []FrameType{FrameWsOpen, FrameWsOpened, FrameWsData, FrameWsClose} {
		if !knownType(frameType) {
			t.Fatalf("frame type %d must be known to the decoder", frameType)
		}
		encoded, err := EncodeFrame(frameType, 1, []byte{0})
		if err != nil {
			t.Fatalf("EncodeFrame(%d): %v", frameType, err)
		}
		frame, err := decodeFrameView(encoded)
		if err != nil {
			t.Fatalf("decodeFrameView(%d): %v", frameType, err)
		}
		if frame.Type != frameType {
			t.Fatalf("round trip gave type %d, want %d", frame.Type, frameType)
		}
	}
	if knownType(FrameWsClose + 1) {
		t.Fatal("the range must not admit types the protocol has not defined")
	}
}

File: packages/protocol/test/websocket.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import {
  FrameType,
  MAX_WS_MESSAGE_BYTES,
  ProtocolError,
  WebSocketDataKind,
  decodeFrame,
  decodeWebSocketClose,
  decodeWebSocketData,
  encodeFrame,
  encodeWebSocketClose,
  encodeWebSocketData,
  isFrameType
} from "../src/index.js";

test("websocket frame types are known", () => {
  assert.equal(isFrameType(FrameType.WsOpen), true);
  assert.equal(isFrameType(FrameType.WsOpened), true);
  assert.equal(isFrameType(FrameType.WsData), true);
  assert.equal(isFrameType(FrameType.WsClose), true);
  assert.equal(isFrameType(15), false);
});

test("a text message survives the round trip as text", () => {
  const body = new TextEncoder().encode("hello wisp");
  const decoded = decodeWebSocketData(encodeWebSocketData(WebSocketDataKind.Text, body));

  assert.equal(decoded.kind, WebSocketDataKind.Text);
  assert.deepEqual([...decoded.data], [...body]);
});

test("a binary message stays binary", () => {
  const body = Uint8Array.from([0, 1, 2, 253, 254, 255]);
  const decoded = decodeWebSocketData(encodeWebSocketData(WebSocketDataKind.Binary, body));

  assert.equal(decoded.kind, WebSocketDataKind.Binary);
  assert.deepEqual([...decoded.data], [...body]);
});

test("an empty message is legal in both kinds", () => {
  for (const kind of [WebSocketDataKind.Text, WebSocketDataKind.Binary]) {
    const decoded = decodeWebSocketData(encodeWebSocketData(kind, new Uint8Array(0)));
    assert.equal(decoded.kind, kind);
    assert.equal(decoded.data.byteLength, 0);
  }
});

test("a message at the ceiling encodes and one past it does not", () => {
  const atLimit = new Uint8Array(MAX_WS_MESSAGE_BYTES);
  assert.equal(encodeWebSocketData(WebSocketDataKind.Binary, atLimit).byteLength, MAX_WS_MESSAGE_BYTES + 1);
  assert.throws(
    () => encodeWebSocketData(WebSocketDataKind.Binary, new Uint8Array(MAX_WS_MESSAGE_BYTES + 1)),
    ProtocolError
  );
});

test("an unknown data kind is rejected rather than guessed", () => {
  assert.throws(() => decodeWebSocketData(Uint8Array.from([7, 1, 2])), ProtocolError);
});

test("an empty data payload is rejected", () => {
  assert.throws(() => decodeWebSocketData(new Uint8Array(0)), ProtocolError);
});

test("a close round trips its code and reason", () => {
  const decoded = decodeWebSocketClose(encodeWebSocketClose(1000, "done"));

  assert.equal(decoded.code, 1000);
  assert.equal(decoded.reason, "done");
});

test("a close with no reason round trips", () => {
  assert.deepEqual(decodeWebSocketClose(encodeWebSocketClose(1001)), { code: 1001, reason: "" });
});

test("the codes an endpoint must synthesise locally never go on the wire", () => {
  for (const code of [1005, 1006]) {
    assert.throws(() => encodeWebSocketClose(code), ProtocolError);
  }
});

test("out-of-range close codes are rejected", () => {
  for (const code of [0, 999, 0x10000]) {
    assert.throws(() => encodeWebSocketClose(code), ProtocolError);
  }
});

test("a close reason longer than the spec allows is rejected", () => {
  assert.throws(() => encodeWebSocketClose(1000, "x".repeat(124)), ProtocolError);
  assert.doesNotThrow(() => encodeWebSocketClose(1000, "x".repeat(123)));
});

test("a truncated close payload is rejected", () => {
  assert.throws(() => decodeWebSocketClose(Uint8Array.from([3])), ProtocolError);
});

test("a websocket message travels inside an ordinary frame", () => {
  const payload = encodeWebSocketData(WebSocketDataKind.Text, new TextEncoder().encode("ping"));
  const frame = decodeFrame(encodeFrame(FrameType.WsData, 7, payload));

  assert.equal(frame.type, FrameType.WsData);
  assert.equal(frame.requestId, 7);
  assert.equal(decodeWebSocketData(frame.payload).kind, WebSocketDataKind.Text);
});
